### PR TITLE
Add apiary functionality

### PR DESCRIPTION
**Pull request by PS-Soundwave**
Migrates bee mating commands to apiary, and implements the life cycle of bees.

Strongly recommended to test alongside SoundyGenetics#3, otherwise drones will stack with potentially genetics erasing effects. The combination of these PRs notably allows drones to stack only if their genetics are the same. Princesses and queens never stack.

The apiary can be interacted with, displaying a UI with two slots arranged vertically on the left, a lifespan bar adjacent to these slots, and a honeycomb of product slots. If a princess is placed in the top left slot and a drone in the bottom left slot, after a time (currently 1 second), indicated by the filling of the lifespan bar in red, they will be mated, with the resulting queen occupying the top left slot.

If a queen occupies the top left slot, it will be active, and after a time being active will have its lifespan tick down, indicated by the emptying of a yellow lifespan bar. Once its life has ticked down completely, it will die, and its offspring, 1 princess and a genetics dependent number of drones, will appear in the output. Additionally, each lifespan tick has a chance of producing a honeycomb in the output.

Testing:
- Ensure that the apiary meets the general description above with no unexpected behavior.
- The top left slot should only accept princesses or queens (not drones or other items).
- The top right slot should only accept drones (not princesses/queens/other items).
- The output slots should only allow the removal of items.
- Closing the UI should not prevent mating or lifespan ticks.
- Changing the item in the princess or drone slot should interrupt mating.
- Placing drones in the bottom left while a queen is in the top left is permitted, but should otherwise have no effect
- Removing a queen and replacing it should not reset its lifespan. (It will however reset the current lifespan tick. This is intended.)
- If there is not a queen present or mating is not occurring, the lifespan bar should be empty.

Advanced Testing:
- Ensure random genetic combination still occurs as expected
- Ensure drone stacking occurs based on genetics
    - Drones spawned using give or from pure bred lines should stack
    - Drones from first-gen cross breeds should stack assuming mutation does not occur
    - Drones from second-gen cross breeds (cross of two first-gen bees) should rarely stack (only 1/16 chance)
    - Obviously, drones of different species should not stack
    - Princesses and queens should not stack
- Ensure that the following traits hold for the three species:
    - Species A - Lifespan: 6 ticks, Tick length: 75 seconds, Production chance: 25% per tick, Drone offspring: 1
    - Species B - Lifespan: 3 ticks, Tick length: 100 seconds, Production chance: 25% per tick, Drone offspring: 2
    - Species C - Lifespan: 9 ticks, Tick length: 50 seconds, Production chance: 50% per tick, Drone offspring: 4

### CODE REVIEW

File: src/main/java/org/terasology/projsndwv/TempBeeRegistry.java [hunk 1]
```diff
@@ -15,26 +15,114 @@
...
+                return null;
+            }
+            return entityManager.create(production.prefab);
+        }
+        else {
+            return EntityRef.NULL;
+        }
+    }
+
+    private static MersenneRandom getRandom() {
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/TempBeeRegistry.java:93]: Is there a particular reason you are trying to tie the random here to the world seed?
Only the world gen itself needs to be deterministic, the gameplay does not.

File: src/main/java/org/terasology/projsndwv/components/ApiaryMatingComponent.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+/***
+ * Indicates to ApiaryScreen that bees are currently mating in an apiary.
+ */
+public final class ApiaryMatingComponent implements Component {
+    public long mateFinishTime;
+
+    private static final Logger logger = LoggerFactory.getLogger(ApiaryMatingComponent.class);
+
+    public ApiaryMatingComponent() {
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/components/ApiaryMatingComponent.java]: This shouldn't be in the component. They should only contain data and the logic in an system

File: src/main/java/org/terasology/projsndwv/components/MatedComponent.java [hunk 3]
```diff
@@ -18,14 +18,42 @@
...
+     * The number of ticks remaining in the lifespan of the queen.
+     */
+    public int ticksRemaining;
+
+    /**
+     * The total ticks in the entire lifespan of this queen.
+     */
+    public int lifespan;
+
+    @SuppressWarnings("unused")
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/components/MatedComponent.java]: At present, the convention is a comment in the javadoc over a suppress warnings.

File: src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java [hunk 4]
```diff
@@ -15,23 +15,273 @@
...
+    }
+
     @ReceiveEvent
-    public void onApiaryInitialized(OnAddedComponent event, EntityRef entity, ApiaryComponent component) {
-        if (!entity.hasComponent(InventoryComponent.class)) {
-            InventoryComponent inventoryComponent = new InventoryComponent(9);
-            inventoryComponent.privateToOwner = false;
-            entity.addComponent(inventoryComponent);
+    public void onApiaryInventoryChanged(InventorySlotChangedEvent event, EntityRef entity, ApiaryComponent component) {
+        DelayManager delayManager = CoreRegistry.get(DelayManager.class);
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java]: You have `@In`ed the delay manager above, why are you fetching it from core registry here?

File: src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java [hunk 5]
```diff
@@ -15,23 +15,273 @@
...
+                        entity.addOrSaveComponent(new ApiaryMatingComponent());
+                        delayManager.addDelayedAction(entity, MATING_EVENT, MATING_TIME);
+                    }
+                }
+            }
+        }
+    }
+
+    @ReceiveEvent
+    public void onApiaryEvent(DelayedActionTriggeredEvent event, EntityRef entity, ApiaryComponent component) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java:199]: I would like to see this split up into smaller functions more. It's a bit too big and monolithic

File: src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java [hunk 6]
```diff
@@ -15,23 +15,273 @@
...
+            BeeComponent beeComponent = femaleBee.getComponent(BeeComponent.class);
+            beeComponent.type = BeeComponent.BeeType.QUEEN;
+            femaleBee.saveComponent(beeComponent);
+            ItemComponent itemComponent = femaleBee.getComponent(ItemComponent.class);
+            itemComponent.icon = TempBeeRegistry.getTextureRegionAssetForSpeciesAndType(femaleBee.getComponent(GeneticsComponent.class).activeGenes.get(ApiarySystem.LOCUS_SPECIES), 2);
+            femaleBee.saveComponent(itemComponent);
+            femaleBee.saveComponent(TempBeeRegistry.getDisplayNameComponentForSpeciesAndType(femaleBee.getComponent(GeneticsComponent.class).activeGenes.get(ApiarySystem.LOCUS_SPECIES), 2));
+            maleBee.destroy();
+
+            DelayManager delayManager = CoreRegistry.get(DelayManager.class);
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java]: My prior comment applies to all instances of this

File: src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java [hunk 7]
```diff
@@ -15,23 +15,273 @@
...
+        }
+    }
+
+    private Genome getGenome() {
+        if (genome == null) {
+            genome = new Genome(4, worldGenerator.getWorldSeed().hashCode());
+
+            GeneticsComponent beeCComponent = new GeneticsComponent(4);
+
+            beeCComponent.activeGenes.add(2); // TODO: (Soundwave) This is bad, both because the size isn't ensured, and because it's messy.
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java:306]: Is this to be addressed in a follow PR or this one?

File: src/main/java/org/terasology/projsndwv/ui/ApiaryScreen.java [hunk 8]
```diff
@@ -88,4 +100,32 @@ public EntityRef get() {
             return entityRef;
         }
     }
+
+    @Override
+    public void update(float delta) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/projsndwv/ui/ApiaryScreen.java:105]: Ideally, this function shouldn't exist. You can use a binding to update the bar, and either a binding or methods that are triggered when needed for the rest

**Review by syntaxi — COMMENTED** [review]
I would like to see javadoc on methods, especially the ones receiving events. You can drop the parameters on those but instead specify what is triggering them.

File: src/main/java/org/terasology/projsndwv/ui/ApiaryScreen.java (hunk 8 above)
Comment (PS-Soundwave) [reply src/main/java/org/terasology/projsndwv/ui/ApiaryScreen.java:105]: Update is necessary for mating. I figured not using it for lifespan as well was unnecessary optimization.

File: src/main/java/org/terasology/projsndwv/TempBeeRegistry.java (hunk 1 above)
Comment (PS-Soundwave) [reply src/main/java/org/terasology/projsndwv/TempBeeRegistry.java:93]: Arbitrary design choice. It's a traditional feature.

File: src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java (hunk 7 above)
Comment (PS-Soundwave) [reply src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java:306]: Actually it's unfortunately potentially out of scope at the moment. Further discussion will be appropriate this weekend.

**Review by Cervator — APPROVED** [review]
Tests out locally! 👍 Or at least I think it did, I was able to do stuff. None of the drones in my batch would stack together, but vanilla ones from the console did. Unsure of the odds there or if something's buggy.

PR is pretty well written, has gone through review (I skimmed it as well), had had improvements applied. Pretty impressed with the turnaround, wish we had seen more of these . Nice job :-)

Included testing the other module PR with this, but will leave any merging for when we have the meeting, all too soon from my still being up at this hour.

**Comment by PS-Soundwave** [comment]
If I get some time before the meeting I'll take a closer look at the stacking to make sure it's functioning correctly. Getting out the next PR is a bit higher priority right now.

**Pull request merged**